### PR TITLE
feat: Integrate Treasure, an NFT Marketplace in Mask Plugin

### LINKED ISSUE #5492: [Bounty] Integrate Treasure, an NFT Marketplace in Mask Plugin
## Catalogue

- Platform
  - [x] Gitcoin
  - [ ] Hackerlink
  - [ ] Mask
- Experience
  - [ ] Beginner
  - [x] Intermediate
  - [ ] Advanced
- Category
  - [ ] Mask Core Feature
  - [x] Mask Plugin
  - [ ] Security/Cryptography Feature
  - [ ] Documentation
  - [ ] Design
  - [ ] Other
    - [ ] /_List it here_/
- Time Commitment
  - [x] 2 weeks
- Bounty Size
  - [x] 150 $Mask or 2000 usd

## Part 1: Problem Description

**Background** 
Treasure is an NFT marketplace on Arbitrum that bridges the growing network of metaverses through an open and composable approach to the convergence of NFTs, DeFi and Gaming.

Cross-ecosystem ties are bolstered through their interrelated resource model; $MAGIC (Power), Treasures (NFT | Resources) and Legions (NFT | Players)

**Solution** 
Create a plugin that integrates LooksRare. The basic product logic could resemble our Opensea and Rarible integration. The UI is in a card style with different tabs serving different purposes.

150004886-c7d03bf0-6099-4310-bb44-c9a8829f2f11

## Part 2: Resources

- [Bounty Guide](https://github.com/DimensionDev/Maskbook/tree/master/docs/bounty-guide.md)
- https://www.treasure.lol/

## Part 3: Contact and Support
@yisiliu @tzhan28 @LindaMask for ideas
@Jack-Works for implementations
Mask Network discord server: https://discord.gg/4Xccj7a75Q,
Treasure discord server: https://discord.com/invite/CzRPER6S5K

Please speak up in the Mask discord's bounty section or just discord dm @yisiliu or @taylor zhang or @lin_dao #5492 so that we can find you on Discord. You can also reach out to lin_DAO on telegram and she will create a dedicated group for faster communication

Submission Requirements
Please include demo to show the integration is working.

Judging Criteria
Have a working demo.
Clean code & tech design
Clear UI, easy to use and generally good user experience

### PR DESCRIPTION
**Pull request by yujqiao**
## Description

This PR integrates Treasure into the Collectible plugin.

I filled out fields that I see Treasure API can provide. However, there are things I'm not sure about and have left blank, like the Listings and the Offers tab. They don't fit very well with Treasure and the History tab should provide all the information. Maybe these two tabs should be hidden with Treasure?

Closes #5492

## Type of change

- [ ] Documentation
- [ ] Code refactoring (Restructuring existing code w/o changing its observable behavior)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (a fix or feature that would make something no longer possible to do/require old user must upgrade their Mask Network to this new version)

## Previews
1. Install the extension
2. Visit https://twitter.com/treasure_sales/status/1487384766018117633 and check out other tweets of the Treasure sales bot.

image
image
image
image
image

**UPDATE:  [hide Offers and Orders when the provider is Treasure](https://github.com/DimensionDev/Maskbook/pull/5688/commits/af06b7961ad4733925fd08986b2d811b2c77254c)**

image

## Checklist

- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
  - [x] I have removed all in development `console.log`s
  - [x] I have removed all commented code.
- [x] I have commented on my code, particularly in hard-to-understand areas.
- [x] I have read [Internationalization Guide](https://github.com/DimensionDev/Maskbook/blob/develop/docs/i18n-guide.md) and moved text fields to the i18n JSON file.

If this PR depends on external APIs:

- [x] I have configured those APIs with CORS headers to let extension requests get passed. 
  - chrome extension: `chrome-extension://[id]`
  - firefox extension: `moz-extension://[id]`
- [x] I have delegated all web requests to the background service via the internal RPC bridge.

### CODE REVIEW

File: packages/web3-providers/src/treasure/index.ts [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+        })
+        if (!response) return
+        return createNFTAsset(response, chainId)
+    }
+
+    async getToken(address: string, tokenId: string) {
+        return createERC721Token(createERC721ContractDetailed(ChainId.Arbitrum, address), {}, tokenId)
+    }
+
+    async getTokens(from: string, opts: NonFungibleTokenAPI.Options) {
```
Comment (yujqiao) [inline packages/web3-providers/src/treasure/index.ts:225]: I haven't find any API I can call for this

**Review by guanbinrui — DISMISSED** [review]

File: packages/mask/src/plugins/Collectible/base.ts [hunk 2]
```diff
@@ -16,6 +16,7 @@ export const base: Plugin.Shared.Definition = {
         postContent: new Set([
             /opensea.io\/assets\/(0x[\dA-Fa-f]{40})\/(\d+)/,
             /rarible.com\/token\/(0x[\dA-Fa-f]{40}):(\d+)/,
+            /marketplace.treasure.lol\/collection\/(0x[\dA-Fa-f]{40})\/(\d+)/,
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Collectible/base.ts]: It does match the following links:

+ https://marketplace.treasure.lol/collection/legion-auxiliary/597
+ https://marketplace.treasure.lol/collection/legion-genesis/2635

Comment (yujqiao) [reply packages/mask/src/plugins/Collectible/base.ts]: Should work now

File: packages/mask/src/plugins/Collectible/SNSAdaptor/Collectible.tsx [hunk 3]
```diff
@@ -129,6 +129,19 @@ export function Collectible(props: CollectibleProps) {
     const chainId = useChainId()
     const { token, asset, provider, tabIndex, setTabIndex } = CollectibleState.useContainer()
 
+    // #region tab controls
+    // Set true to hide the corresponding tabs
+    const tabControl = {
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Collectible/SNSAdaptor/Collectible.tsx]: Plz move this to `packages/mask/src/plugins/Collectible/constants.ts` and add record type.

```ts
const TAB_CONTROL: Record<NonFungibleAssetProvider, {
    offers?: boolean
    listing?: boolean
}>
```

**Review by guanbinrui — DISMISSED** [review]

**Review by guanbinrui — DISMISSED** [review]

File: packages/web3-shared/evm/pipes/index.ts [hunk 4]
```diff
@@ -232,6 +236,8 @@ export function resolveCollectibleLink(
             return ''
         case NonFungibleAssetProvider.NFTSCAN:
             return ''
+        case NonFungibleAssetProvider.TREASURE:
+            return ''
```
Comment (guanbinrui) [inline packages/web3-shared/evm/pipes/index.ts:249]: Can we compose a link like this?

`https://marketplace.treasure.lol/collection/legion-genesis/2036`

File: packages/web3-providers/src/treasure/utils.ts [hunk 5]
```diff
@@ -0,0 +1,15 @@
+import BigNumber from 'bignumber.js'
+import { isOne, pow10 } from '@masknet/web3-shared-base'
+
+export function getOrderUnitPrice(currentPrice?: string, decimals?: number, quantity?: string) {
+    if (!currentPrice || !decimals || !quantity) return
+    const _currentPrice = new BigNumber(currentPrice).div(pow10(decimals))
```
Comment (guanbinrui) [inline packages/web3-providers/src/treasure/utils.ts:6]: Recommend to use `dividedBy` and `multipliedBy`.

File: packages/web3-providers/src/treasure/index.ts [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+        if (!this.collectionsCache) {
+            const response = await fetchCollections()
+            if (response !== undefined) this.collectionsCache = response
+        }
+
+        if (!this.collectionsCache) return
+
+        return this.collectionsCache.collections
+            .filter((e) => e.name.toLowerCase().replace(' ', '-') === collectionName)
+            .map((e) => e.id)[0]
```
Comment (guanbinrui) [inline packages/web3-providers/src/treasure/index.ts]:
```suggestion
        return first(this.collectionsCache.collections
            .filter((e) => e.name.toLowerCase().replace(' ', '-') === collectionName)
            .map((e) => e.id))
```

File: packages/web3-providers/src/treasure/index.ts [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+                          link: makeTreasureLink(asset.collection.id, asset.collection.tokens[0].tokenId),
+                      }
+                    : undefined,
+            },
+            price: {
+                quantity: Number(e.pricePerItem) / 1e18 + ' $MAGIC',
+                price: Number(e.pricePerItem) / 1e18 + ' $MAGIC',
+            },
+            eventType: e.status,
+            timestamp: Number(e.blockTimestamp) * 1000,
```
Comment (guanbinrui) [inline packages/web3-providers/src/treasure/index.ts:175]: Recommend to use `fromUnixTime`.

File: packages/web3-providers/src/treasure/index.ts [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+                to: e.buyer
+                    ? {
+                          address: e.buyer.id,
+                          link: makeTreasureLink(asset.collection.id, asset.collection.tokens[0].tokenId),
+                      }
+                    : undefined,
+            },
+            price: {
+                quantity: Number(e.pricePerItem) / 1e18 + ' $MAGIC',
+                price: Number(e.pricePerItem) / 1e18 + ' $MAGIC',
```
Comment (guanbinrui) [inline packages/web3-providers/src/treasure/index.ts:172]: Recommend to use `<FormattedBalance />` in the UI.

File: packages/web3-providers/src/treasure/index.ts [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+
+function makeTreasureLink(collectionId: string, tokenId?: string) {
+    if (tokenId !== null) {
+        return `https://marketplace.treasure.lol/collection/${collectionId}/${tokenId}`
+    } else {
+        return `https://marketplace.treasure.lol/collection/${collectionId}/`
+    }
+}
+
+function resolveImageUrl(url: string): string {
```
Comment (guanbinrui) [inline packages/web3-providers/src/treasure/index.ts:94]: Replace it with `resolveResourceLink`.

**Comment by guanbinrui** [comment]
@rapiz1 Any updates?

**Pull request closed without merging**